### PR TITLE
implement openshiftFormatter

### PR DESCRIPTION
**Pull request by livetocode**
## Hey, I just made a Pull Request!

implement function openshiftFormatter, as well as the unit tests.

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: .changeset/green-bags-hug.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-kubernetes': minor
```
Comment (jhaals) [inline .changeset/green-bags-hug.md:2]: I believe this should be a patch since we are below version 1 and it should be breaking(?)

**Review by jhaals — COMMENTED** [review]
Nice work implementing this 👍

File: .changeset/green-bags-hug.md (hunk 1 above)
Comment (Xantier) [reply .changeset/green-bags-hug.md:2]: Yup, patch is the new minor in sub 1 version.

*not breaking, I assume. Looks to be additional

Comment (livetocode) [reply .changeset/green-bags-hug.md]: Done!

File: plugins/kubernetes/src/utils/clusterLinks/formatters/openshift.ts [hunk 2]
```diff
@@ -15,10 +15,30 @@
...
+};
+
+export function openshiftFormatter(options: ClusterLinksFormatterOptions): URL {
+  const result = new URL(options.dashboardUrl.href);
+  const name = options.object.metadata?.name;
+  const namespace = options.object.metadata?.namespace;
+  const validKind = kindMappings[options.kind.toLocaleLowerCase('en-US')];
+  if (namespace) {
+    if (name && validKind) {
+      result.pathname = `k8s/ns/${namespace}/${validKind}/${name}`;
```
Comment (freben) [inline plugins/kubernetes/src/utils/clusterLinks/formatters/openshift.ts:33]: Would you mind adding an `encodeURIComponent(x)` on all dynamic path parts? I know that it's common to only permit URL safe chars for these parts, but Backstage permits users to tweak those rules. And either way - it's good practice to not trust user input :)

Comment (freben) [reply plugins/kubernetes/src/utils/clusterLinks/formatters/openshift.ts:33]: Not if they contain slashes ;) plus that auto quoting behavior is really odd - it tries its best to avoid double encoding etc using some heuristic 🤔

Comment (livetocode) [reply plugins/kubernetes/src/utils/clusterLinks/formatters/openshift.ts:33]: ok, fine, I will do the encoding in all formatters...
I thought that forcing the URL type would solve the issue of ignoring the encoding, but if it doesn't, I might revert to the String type...

Comment (livetocode) [reply plugins/kubernetes/src/utils/clusterLinks/formatters/openshift.ts]: I did the refactoring and the encoding

File: plugins/kubernetes/src/types/types.ts [hunk 3]
```diff
@@ -46,4 +46,4 @@ export interface ClusterLinksFormatterOptions {
 
 export type ClusterLinksFormatter = (
   options: ClusterLinksFormatterOptions,
-) => URL;
+) => string;
```
Comment (freben) [inline plugins/kubernetes/src/types/types.ts:49]: Why did you change this contract for all formatters? Seems like a separate discussion to take.

File: plugins/kubernetes/src/utils/clusterLinks/formatters/openshift.test.ts [hunk 4]
```diff
@@ -15,21 +15,98 @@
  */
 import { openshiftFormatter } from './openshift';
 
-describe('clusterLinks - OpenShift formatter', () => {
+describe('clusterLinks - Openshift formatter', () => {
```
Comment (freben) [inline plugins/kubernetes/src/utils/clusterLinks/formatters/openshift.test.ts:18]: they actually capitalize it `OpenShift`, so we should probably keep it that way here and in the vocab file

File: plugins/kubernetes/src/utils/clusterLinks/formatters/openshift.ts [hunk 5]
```diff
@@ -15,10 +15,39 @@
...
+    }
+  } else if (validKind) {
+    segments.push('cluster', validKind);
+    if (name) {
+      segments.push(name);
+    }
+  }
+  if (segments.length > 1) {
+    // define a path only when we have more segments than the initial 'k8s'
+    result.pathname = segments
```
Comment (freben) [inline plugins/kubernetes/src/utils/clusterLinks/formatters/openshift.ts:48]: So what happens here is you replace the entire path unconditionally. What if your base URL is `https://example.com/x/openshift-web`? Seems to me that you should have kept the URL return type, and used the relative resolution constructor of URL.

```ts
const name = encodeURIComponent(options.object.metadata?.name ?? '');
const namespace = encodeURIComponent(options.object.metadata?.namespace ?? '');
const validKind = kindMappings[options.kind.toLocaleLowerCase('en-US')];

// this works well with the constructor on the last line, so you can also
// skip the segments.length === 1 check
let path = ''; 
if (namespace) {
  if (name && validKind) {
    path = `k8s/ns/${namespace}/${validKind}/${name}`;
  }
  // ...etc

return new URL(path, options.dashboardUrl.href);
```

Side note: maybe the validKinds thing could be replaced with a `pluralizeKind` function that just returns the input + `'s'` except in special circumstances? :P That would make it work magically for most english words without extra effort. And then that could be made into a helper function that's shared for the entire project later on

Comment (livetocode) [reply plugins/kubernetes/src/utils/clusterLinks/formatters/openshift.ts]: I rolled back my changes and refactored the code with your suggestions.
I applied them to the other formatters and I moved the special case of URL rendering into the standardFormatter which needs url fragments, but not the other ones.
I think this is cleaner that way.

File: .changeset/green-bags-hug.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-kubernetes': patch
+---
+
+Implement support for formatting Openshift dashboard url links
```
Comment (freben) [inline .changeset/green-bags-hug.md:5]: OK so now Vale says (correctly!) that this should be OpenShift with a capital S :)

**Review by freben — COMMENTED** [review]
aaaalmost there!

File: .changeset/green-bags-hug.md (hunk 6 above)
Comment (livetocode) [reply .changeset/green-bags-hug.md]: Oups! Fixed.

**Review by freben — APPROVED** [review]
